File: src/weapons.ts
// === Weapon definitions — spread bloom + recoil profiles ===

export type WeaponType = 'Pistol' | 'Shotgun' | 'SMG' | 'Sniper' | 'AR';

export interface WeaponState {
  type: WeaponType;
  name: string;
  damage: number;
  fireRate: number;
  bulletSpeed: number;
  bulletCount: number;
  /** Base cone in degrees (half-width random for singles, full fan for multi) */
  baseSpread: number;
  /** Max bloom degrees added from sustained fire */
  maxBloom: number;
  /** Degrees of bloom added per shot */
  bloomPerShot: number;
  /** Bloom recovery deg/sec */
  bloomRecover: number;
  /** Instant aim kick per shot (deg), recovers smoothly */
  recoilKick: number;
  /** Recoil recover deg/sec */
  recoilRecover: number;
  /** Camera punch 0–1 */
  punch: number;
  ammo: number | null;
  maxAmmo: number | null;
  magSize: number | null;
  mag: number | null;
  reloadTime: number;
  reloadCooldown: number;
  fireCooldown: number;
  range: number;
  penetrate: boolean;
  /** Runtime bloom heat */
  bloom: number;
}

export const LOADOUT_OPTIONS: WeaponType[] = ['SMG', 'AR', 'Shotgun', 'Sniper'];

export function createWeapon(type: WeaponType): WeaponState {
  switch (type) {
    case 'Pistol':
      return {
        type, name: 'Pistol', damage: 16, fireRate: 4.0, bulletSpeed: 680,
        bulletCount: 1, baseSpread: 0.8, maxBloom: 4.5, bloomPerShot: 1.4, bloomRecover: 10,
        recoilKick: 1.6, recoilRecover: 14, punch: 0.12,
        ammo: null, maxAmmo: null, magSize: 12, mag: 12,
        reloadTime: 1.1, reloadCooldown: 0, fireCooldown: 0, range: 540, penetrate: false, bloom: 0,
      };
    case 'Shotgun':
      return {
        type, name: 'Shotgun', damage: 8, fireRate: 1.05, bulletSpeed: 540,
        bulletCount: 8, baseSpread: 22, maxBloom: 8, bloomPerShot: 3, bloomRecover: 6,
        recoilKick: 5.5, recoilRecover: 9, punch: 0.45,
        ammo: 24, maxAmmo: 24, magSize: 6, mag: 6,
        reloadTime: 1.7, reloadCooldown: 0, fireCooldown: 0, range: 230, penetrate: false, bloom: 0,
      };
    case 'SMG':
      return {
        type, name: 'SMG', damage: 7, fireRate: 12.5, bulletSpeed: 720,
        bulletCount: 1, baseSpread: 1.2, maxBloom: 11, bloomPerShot: 0.85, bloomRecover: 16,
        recoilKick: 1.1, recoilRecover: 18, punch: 0.08,
        ammo: 120, maxAmmo: 120, magSize: 30, mag: 30,
        reloadTime: 1.65, reloadCooldown: 0, fireCooldown: 0, range: 400, penetrate: false, bloom: 0,
      };
    case 'AR':
      return {
        type, name: 'AR', damage: 12, fireRate: 7.2, bulletSpeed: 820,
        bulletCount: 1, baseSpread: 0.6, maxBloom: 7, bloomPerShot: 0.7, bloomRecover: 12,
        recoilKick: 1.9, recoilRecover: 15, punch: 0.18,
        ammo: 90, maxAmmo: 90, magSize: 25, mag: 25,
        reloadTime: 1.9, reloadCooldown: 0, fireCooldown: 0, range: 580, penetrate: false, bloom: 0,
      };
    case 'Sniper':
      return {
        type, name: 'Sniper', damage: 78, fireRate: 0.7, bulletSpeed: 1200,
        bulletCount: 1, baseSpread: 0.15, maxBloom: 2, bloomPerShot: 2, bloomRecover: 4,
        recoilKick: 8, recoilRecover: 7, punch: 0.55,
        ammo: 12, maxAmmo: 12, magSize: 4, mag: 4,
        reloadTime: 2.3, reloadCooldown: 0, fireCooldown: 0, range: 920, penetrate: true, bloom: 0,
      };
  }
}

export function ammoDisplay(w: WeaponState): string {
  if (w.reloadCooldown > 0) return '…';
  if (w.magSize !== null && w.mag !== null) {
    if (w.ammo === null) return `${w.mag}`;
    return `${w.mag}/${w.ammo}`;
  }
  if (w.ammo === null) return '∞';
  return `${w.ammo}`;
}

export function loadoutLabel(type: WeaponType): string {
  switch (type) {
    case 'SMG': return 'SMG — spray control';
    case 'AR': return 'AR — mid range';
    case 'Shotgun': return 'Shotgun — close';
    case 'Sniper': return 'Sniper — one-tap';
    default: return type;
  }
}

File: src-tauri/src/game/weapons.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub enum WeaponType {
    Pistol,
    Shotgun,
    SMG,
    AR,
    Sniper,
}

impl WeaponType {
    pub fn from_str_loose(s: &str) -> Self {
        match s.to_ascii_lowercase().as_str() {
            "smg" => WeaponType::SMG,
            "ar" | "rifle" => WeaponType::AR,
            "shotgun" => WeaponType::Shotgun,
            "sniper" => WeaponType::Sniper,
            "pistol" => WeaponType::Pistol,
            _ => WeaponType::AR,
        }
    }

    pub fn as_key(self) -> &'static str {
        match self {
            WeaponType::Pistol => "Pistol",
            WeaponType::Shotgun => "Shotgun",
            WeaponType::SMG => "SMG",
            WeaponType::AR => "AR",
            WeaponType::Sniper => "Sniper",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Weapon {
    pub weapon_type: WeaponType,
    pub name: String,
    pub damage: i32,
    pub fire_rate: f64,
    pub bullet_speed: f64,
    pub bullet_count: u32,
    pub spread: f64,
    pub ammo: Option<i32>,
    pub max_ammo: Option<i32>,
    pub reload_time: f64,
    pub reload_cooldown: f64,
    pub fire_cooldown: f64,
    pub range: f64,
    pub penetrate: bool,
}

impl Weapon {
    pub fn new(weapon_type: WeaponType) -> Self {
        match weapon_type {
            WeaponType::Pistol => Self {
                weapon_type,
                name: "Pistol".into(),
                damage: 14,
                fire_rate: 3.2,
                bullet_speed: 620.0,
                bullet_count: 1,
                spread: 2.0,
                ammo: None,
                max_ammo: None,
                reload_time: 0.0,
                reload_cooldown: 0.0,
                fire_cooldown: 0.0,
                range: 520.0,
                penetrate: false,
            },
            WeaponType::Shotgun => Self {
                weapon_type,
                name: "Shotgun".into(),
                damage: 9,
                fire_rate: 0.95,
                bullet_speed: 520.0,
                bullet_count: 7,
                spread: 28.0,
                ammo: Some(24),
                max_ammo: Some(24),
                reload_time: 1.6,
                reload_cooldown: 0.0,
                fire_cooldown: 0.0,
                range: 240.0,
                penetrate: false,
            },
            WeaponType::SMG => Self {
                weapon_type,
                name: "SMG".into(),
                damage: 7,
                fire_rate: 11.0,
                bullet_speed: 700.0,
                bullet_count: 1,
                spread: 7.0,
                ammo: Some(120),
                max_ammo: Some(120),
                reload_time: 1.8,
                reload_cooldown: 0.0,
                fire_cooldown: 0.0,
                range: 420.0,
                penetrate: false,
            },
            WeaponType::AR => Self {
                weapon_type,
                name: "AR".into(),
                damage: 12,
                fire_rate: 6.5,
                bullet_speed: 780.0,
                bullet_count: 1,
                spread: 3.0,
                ammo: Some(90),
                max_ammo: Some(90),
                reload_time: 2.0,
                reload_cooldown: 0.0,
                fire_cooldown: 0.0,
                range: 560.0,
                penetrate: false,
            },
            WeaponType::Sniper => Self {
                weapon_type,
                name: "Sniper".into(),
                damage: 72,
                fire_rate: 0.55,
                bullet_speed: 1100.0,
                bullet_count: 1,
                spread: 0.0,
                ammo: Some(12),
                max_ammo: Some(12),
                reload_time: 2.4,
                reload_cooldown: 0.0,
                fire_cooldown: 0.0,
                range: 900.0,
                penetrate: true,
            },
        }
    }

    pub fn can_fire(&self) -> bool {
        if self.fire_cooldown > 0.0 || self.reload_cooldown > 0.0 {
            return false;
        }
        match self.ammo {
            Some(ammo) => ammo > 0,
            None => true,
        }
    }

    pub fn fire(&mut self) {
        self.fire_cooldown = 1.0 / self.fire_rate;
        if let Some(ref mut ammo) = self.ammo {
            *ammo -= 1;
        }
    }

    pub fn start_reload(&mut self) -> bool {
        if self.reload_cooldown > 0.0 {
            return false;
        }
        if let (Some(ammo), Some(max_ammo)) = (self.ammo, self.max_ammo) {
            if ammo >= max_ammo {
                return false;
            }
        } else {
            return false;
        }
        self.reload_cooldown = self.reload_time;
        true
    }

    pub fn update_cooldowns(&mut self, delta_time: f64) {
        if self.fire_cooldown > 0.0 {
            self.fire_cooldown = (self.fire_cooldown - delta_time).max(0.0);
        }
        if self.reload_cooldown > 0.0 {
            self.reload_cooldown -= delta_time;
            if self.reload_cooldown <= 0.0 {
                self.reload_cooldown = 0.0;
                if let Some(ref mut ammo) = self.ammo {
                    if let Some(max_ammo) = self.max_ammo {
                        *ammo = max_ammo;
                    }
                }
            }
        }
    }

    pub fn ammo_display(&self) -> String {
        if self.reload_cooldown > 0.0 {
            return "…".into();
        }
        match self.ammo {
            Some(ammo) => {
                if let Some(max_ammo) = self.max_ammo {
                    format!("{}/{}", ammo, max_ammo)
                } else {
                    format!("{}", ammo)
                }
            }
            None => "∞".into(),
        }
    }
}

pub fn create_pickup_weapon(weapon_type: WeaponType) -> Weapon {
    Weapon::new(weapon_type)
}
